src/csharp/Grpc.Core/Internal/CallSafeHandle.cs

```diff
@@ -134,9 +135,48 @@ public void StartSendCloseFromClient(ISendCompletionCallback callback)
             {
                 var ctx = completionQueue.CompletionRegistry.RegisterBatchCompletion(CompletionHandler_ISendStatusFromServerCompletionCallback, callback);
                 var optionalPayloadLength = optionalPayload != null ? new UIntPtr((ulong)optionalPayload.Length) : UIntPtr.Zero;
-                var statusDetailBytes = MarshalUtils.GetBytesUTF8(status.Detail);
-                Native.grpcsharp_call_send_status_from_server(this, ctx, status.StatusCode, statusDetailBytes, new UIntPtr((ulong)statusDetailBytes.Length), metadataArray, sendEmptyInitialMetadata ? 1 : 0,
-                    optionalPayload, optionalPayloadLength, writeFlags).CheckOk();
+
+                const int MaxStackAllocBytes = 256;
+                int maxBytes = MarshalUtils.GetMaxByteCountUTF8(status.Detail);
+                if (maxBytes > MaxStackAllocBytes)
+                {
+                    // pay the extra to get the *actual* size; this could mean that
+                    // it ends up fitting on the stack after all, but even if not
+                    // it will mean that we ask for a *much* smaller buffer
+                    maxBytes = MarshalUtils.GetByteCountUTF8(status.Detail);
+                }
+
+                unsafe
```

Comment:
btw, there might be more where a string is passed to the native layer and that could benefit from this trick. I think the idea used in this PR is good, but we should figure out how to extract the logic to a utility function and reuse it in multiple places as a followup.